main.go

```diff
@@ -101,6 +91,70 @@ func main() {
 		puzzle.Board = append(puzzle.Board, cells)
 	}
 
+	return puzzle
+}
+
+func main() {
+	cmdSolve := &cobra.Command{
+		Use:   "solve [puzzle file]",
+		Short: "Solve the given Sudoku puzzle",
+		Long:  `A sudoku puzzle given to stdin will be validated and solved`,
+		Run: func(cmd *cobra.Command, args []string) {
+			var reader io.Reader
+			if len(args) > 0 {
+				// read the puzzle from the given file
+				file, err := os.Open(args[0])
+				if err != nil {
+					fmt.Printf("Unable to read file: %s\n", args[0])
+					os.Exit(1)
+				}
+
+				reader = file
+			} else {
+				file, err := os.Stdin.Stat()
+				if err != nil {
+					fmt.Printf("Error checking stdin: %v\n", err)
+					os.Exit(1)
+				}
+				waitingForUserInput := (file.Mode() & os.ModeCharDevice) != 0
```

Comment:
Detect if a file (with a puzzle) has been piped/redirect to the program or if we are waiting on user input.

`os.ModeCharDevice` is a file mode that indicates the device is waiting for character by character input from the terminal rather than piped input.